torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1238,34 +1276,44 @@ def _optim_state_dict(
     )
     fqn_to_fsdp_param_info = _get_fqn_to_fsdp_param_info(model)
 
-    (
-        param_id_to_optim_state_key,
-        optim_state_key_to_param_id,
-    ) = _map_param_id_to_optim_keys(
+    all_optim_state_keys, optim_state_key_to_param_id = _map_param_id_to_optim_keys(
         optim_state_dict,
         group,
         param_id_to_param,
         param_to_fqns,
         fqn_to_fsdp_param_info,
+        merge_keys=use_orig_params,
     )
 
     # Iterate in rank 0's flattened parameter ID order to ensure aligned
     # all-gathers across ranks
-    for optim_state_key in param_id_to_optim_state_key.values():
-        param_id = optim_state_key_to_param_id[optim_state_key]
+    for optim_state_key in all_optim_state_keys:
+        param_id = optim_state_key_to_param_id.get(optim_state_key, -1)
+        assert param_id >= 0 or (optim_state_key.is_fsdp_managed and use_orig_params)
```

Comment:
nit: Can we add a message to the `assert` explaining that for `use_orig_params=True`, some parameters may not be in the local shard, so they will be missing the optimizer state keys?